### PR TITLE
Hugging Face Transformer Deployment Tutorial

### PR DESCRIPTION
**Pull request by fpetrini15**
Tutorials to show how hugging face transformers can be quickly deployed in Triton.

### CODE REVIEW

File: Quick_Deploy/HuggingFaceTransformers/base_text_classification_model.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            # Assume input named "prompt" for now, specified in autocomplete above
+            prompt_tensor = pb_utils.get_input_tensor_by_name(request, "prompt")
+            prompt = prompt_tensor.as_numpy()[0].decode("utf-8")
+
+            self.logger.log_info(f"Conducting sentiment analysis of prompt: {prompt}")
+            response = self.generate(prompt)
+            responses.append(response)
+        return responses
+
+    def generate(self, prompt):
```
Comment (nnshah1) [inline Quick_Deploy/HuggingFaceTransformers/base_text_classification_model.py:78]:
```suggestion
    def classify(self, prompt):
```

File: Quick_Deploy/HuggingFaceTransformers/base_text_classification_model.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        [cfg.add_input(i) for i in inputs]
+        [cfg.add_output(o) for o in outputs]
+        cfg.set_max_batch_size(0)
+        return cfg
+
+    def initialize(self, args):
+        self.logger = pb_utils.Logger
+        self.model_config = json.loads(args["model_config"])
+        self.model_params = self.model_config.get("parameters", {})
+        default_hf_model = "facebook/opt-125m"
```
Comment (nnshah1) [inline Quick_Deploy/HuggingFaceTransformers/base_text_classification_model.py:53]: question: is the same default model for generation and classification?

File: Quick_Deploy/HuggingFaceTransformers/build_server.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+#    documentation and/or other materials provided with the distribution.
+#  * Neither the name of NVIDIA CORPORATION nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
```
Comment (nnshah1) [inline Quick_Deploy/HuggingFaceTransformers/build_server.py:19]: could we have a docker file with build arguments for the tutorial instead of generating one via script?

I can see a separate script for 'generate_mode_repository.py' that would take in list of hf models and build out the model repository

then docker run script to just run.

Guess I'm thinking tutorial wise - may be easier to grok as seperate with docker build and docker run being bash.

Theoretically:

create_repository.py
Dockerfile
build.sh
run.sh

File: Quick_Deploy/HuggingFaceTransformers/base_text_generation_model.py
Comment (nnshah1) [inline Quick_Deploy/HuggingFaceTransformers/base_text_generation_model.py:1]: suggestion: perhaps to rename as:
```
text_classification/model.py
text_generation/model.py
```

File: Quick_Deploy/HuggingFaceTransformers/base_text_classification_model.py (hunk 2 above)
Comment (fpetrini15) [reply Quick_Deploy/HuggingFaceTransformers/base_text_classification_model.py:53]: Thank you for catching. Yes, and they should not be.

File: Quick_Deploy/HuggingFaceTransformers/build_server.py (hunk 3 above)
Comment (fpetrini15) [reply Quick_Deploy/HuggingFaceTransformers/build_server.py:19]: We can definitely do it this way. My reasoning to have everything build in a single script was based on "one-click-deployment". I was hoping a single script would abstract users from all the files and details that go into running Triton. If they wanted to look deeper into the deployment mechanisms, the build script wouldn't delete any of the files it created unless there is a re-run.

I can change the script to accept a list of hf models. I'll try to find some more general models since some require special handling and won't be compatible with our "base model" scripts.

Since the docker command will be one-liners, are we okay following the same format as the [vllm](https://github.com/triton-inference-server/tutorials/tree/main/Quick_Deploy/vLLM#step-1-build-a-triton-container-image-with-vllm) tutorial as opposed to creating sh files?

File: Quick_Deploy/HuggingFaceTransformers/create_repository.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+import os
+import argparse
+import json
+import shutil
+import subprocess
+from transformers import pipeline, AutoTokenizer
+
+
+class ServerBuilder:
```
Comment (rmccorm4) [inline Quick_Deploy/HuggingFaceTransformers/create_repository.py:34]: I'm not sure how I feel about this builder script. Part of me would rather focus on making the huggingface a backend or just simply intuitive to use.

For example, the config files should have sensible default values for model name and max length so users can run it as-is, but also see intuitive places to replace the values with their own values. But if you do that, then this script gets harder to write because it's harder to find/replace the "placeholder" tags you put.

Comment (rmccorm4) [reply Quick_Deploy/HuggingFaceTransformers/create_repository.py:34]: If this script were to stay, I'd rather see the default values in `config.pbtxt` be actually usable values, and make the folders here in the repo actually in correct model repository structure by default.

If you want to generate a config file with a CLI like this, you could just overwrite the config file from a template rather than trying to find/replace.

Comment (rmccorm4) [reply Quick_Deploy/HuggingFaceTransformers/create_repository.py:34]: At a minimum, I would lean towards separating this PR into 2. I think we should add the huggingface example model with sensible default values in config and files in model_repository structure that can just be run as-is.

Separately, can discuss a "builder" script or other deployment tool along those lines. I wouldn't bundle them here for now.

File: Quick_Deploy/HuggingFaceTransformers/text_generation/config.pbtxt [hunk 5]
```diff
@@ -0,0 +1,24 @@
+backend: "python"
+parameters: {
+	key: "huggingface_model",
+	value: {string_value: "MODEL_TAG"}
```
Comment (rmccorm4) [inline Quick_Deploy/HuggingFaceTransformers/text_generation/config.pbtxt:4]: use real default value

File: Quick_Deploy/HuggingFaceTransformers/text_generation/config.pbtxt [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+}
+parameters: {
+	key: "max_length",
+	value: {string_value: "MAX_LENGTH_TAG"}
```
Comment (rmccorm4) [inline Quick_Deploy/HuggingFaceTransformers/text_generation/config.pbtxt:8]: use real default value

File: Quick_Deploy/HuggingFaceTransformers/README.md [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+Without specifying any arguments, this script will create a model repository containing
+the details necessary for Triton to load and serve each of the aforementioned models.
+For the purposes of this tutorial, we provide Triton with simple ```config.pbtxt``` files
+for each model that describe:
+- The the backend to use.
+- Model input and output details.
+- Custom parameters to use for execution.
+
+## Step 2: Build a Triton Container Image
```
Comment (yeahdongcn) [inline Quick_Deploy/HuggingFaceTransformers/README.md:62]: I'm confused about the best practice in shipping a custom Python execution environment, spawning a new image, or using [conda-pack](https://github.com/triton-inference-server/python_backend#creating-custom-execution-environments).

What do you prefer in production-level deployment?

File: Quick_Deploy/HuggingFaceTransformers/create_repository.py (hunk 4 above)
Comment (nnshah1) [reply Quick_Deploy/HuggingFaceTransformers/create_repository.py:34]: Agree, directionally the tutorial should be easy to follow / customize / few steps - but shouldn't hide too much w.r.t to automation - I think that will come when we add support into either additional tools or backends.

File: Quick_Deploy/HuggingFaceTransformers/README.md [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+I0922 23:28:40.352017 1 http_server.cc:3558] Started HTTPService at 0.0.0.0:8000
+I0922 23:28:40.395611 1 http_server.cc:187] Started Metrics Service at 0.0.0.0:8002
+```
+
+## Step 4: Use a Triton Client to Query the Server
+
+Now we can query the server using the client script provided.
+
+```bash
+python3 client.py -m [model_name] --prompt [prompt_message]
```
Comment (nnshah1) [inline Quick_Deploy/HuggingFaceTransformers/README.md:94]: can we add curl as reference?

File: Quick_Deploy/HuggingFaceTransformers/README.md (hunk 7 above)
Comment (nnshah1) [reply Quick_Deploy/HuggingFaceTransformers/README.md:62]: @Tabrizian  - thoughts on this?

File: Quick_Deploy/HuggingFaceTransformers/build_server.py (hunk 3 above)
Comment (nnshah1) [reply Quick_Deploy/HuggingFaceTransformers/build_server.py]: Nice work! docker one liners fine and preferred in many scenarios.

File: Quick_Deploy/HuggingFaceTransformers/README.md (hunk 8 above)
Comment (rmccorm4) [reply Quick_Deploy/HuggingFaceTransformers/README.md:94]: This would be blocked until Guan and I finish implementing simpler curl format support, unless you mean a tensor-oriented curl:
```
curl -X POST localhost:8000/v2/models/text/infer -d '{"inputs": [{"name":"prompt","datatype":"BYTES","shape":[1,1],"data":["hello"]}]}'
```

Comment (nnshah1) [reply Quick_Deploy/HuggingFaceTransformers/README.md:94]: I think the simple curl one liner is good for now - and we will add the generate version when available

File: Quick_Deploy/HuggingFaceTransformers/README.md (hunk 7 above)
Comment (dyastremsky) [reply Quick_Deploy/HuggingFaceTransformers/README.md:62]: That depends on your goals. If you want to run models with different Python environments on one server, you would [use custom execution environments](https://github.com/triton-inference-server/python_backend#creating-custom-execution-environments). It's more memory-intensive (especially as you increase the number of models), so if you only need one Python version and environment, it would be easier to use an image with all the components needed.

Also, conda-pack can reduce the size of the environment itself (versus using an uncompressed environment), but it then needs to be unpacked at load time, which increases the model load time considerably.

Comment (Tabrizian) [reply Quick_Deploy/HuggingFaceTransformers/README.md:62]: Adding to what David mentioned, if you don't want to create a new container that has all the dependencies or you want to isolate the dependencies used by each model custom execution environments are the recommended path.

File: Quick_Deploy/HuggingFaceTransformers/README.md [hunk 9]
```diff
@@ -26,125 +26,155 @@
...
+"Where are you going? I’m going to the beach."
 
-```bash
-python3 client.py -m opt-125m --prompt "Who am I?"
-python3 client.py -m gpt2 --prompt "Who am I?"
+# persimmon8b
+"Where is the nearest starbucks?"
 ```
-Which will return a result similar to:
+## 'Hour Zero' Support
```
Comment (rmccorm4) [inline Quick_Deploy/HuggingFaceTransformers/README.md:153]: "Hour zero" was just me being cheeky. I think we should probably call it "day zero" or something more generic like "bleeding edge models" or something. CC @nnshah1 who probably has some thoughts

File: Quick_Deploy/HuggingFaceTransformers/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+FROM nvcr.io/nvidia/tritonserver:23.09-py3
+RUN pip install transformers==4.34.0
+RUN pip install protobuf==3.20.3
+RUN pip install sentencepiece==0.1.99
+RUN pip install accelerate==0.23.0
+RUN pip install einops==0.6.1
```
Comment (rmccorm4) [inline Quick_Deploy/HuggingFaceTransformers/Dockerfile:31]: I think docker best practice is to combine similar build/install instructions like this into a single layer
```suggestion
RUN pip install transformers==4.34.0 protobuf==3.20.3 sentencepiece==0.1.99 accelerate==0.23.0 einops==0.6.1
```

File: Quick_Deploy/HuggingFaceTransformers/README.md [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+models will be deployed:
+- [tiiuae/falcon-7b](https://huggingface.co/tiiuae/falcon-7b)
+- [adept/persimmon-8b-base](https://huggingface.co/adept/persimmon-8b-base)
+- [mistralai/Mistral-7B-v0.1](https://huggingface.co/mistralai/Mistral-7B-v0.1)
+
+These models were selected because of their popularity and consistent response quality.
+However, this tutorial is also generalizable for any transformer model provided 
+sufficient infrastructure. 
+
+*NOTE*: The tutorial is intended to be a reference example only. It is a work in progress.
```
Comment (rmccorm4) [inline Quick_Deploy/HuggingFaceTransformers/README.md:42]:
```suggestion
*NOTE*: The tutorial is intended to be a reference example only. It may not be tuned for optimal performance.
```

Comment (rmccorm4) [reply Quick_Deploy/HuggingFaceTransformers/README.md:42]: Calling it a WIP may deter users from trying it, thinking it may not work.

File: Quick_Deploy/HuggingFaceTransformers/README.md [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+
+## Step 1: Create a Model Repository
+
+The first step is to create a model repository containing the models we want the Triton 
+Inference Server to load and use for inference processing. For this tutorial an empty
+directory named `model_repository` has been provided, into which we will copy the 
+`falcon7b` model folder:
```
Comment (rmccorm4) [inline Quick_Deploy/HuggingFaceTransformers/README.md:49]: Empty folder won't work on git. Just update instructions to `mkdir -p model_repository` first.

File: Quick_Deploy/HuggingFaceTransformers/README.md [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+
+The server has launched successfully when you see the following outputs in your console:
+
+```
+I0922 23:28:40.351809 1 grpc_server.cc:2451] Started GRPCInferenceService at 0.0.0.0:8001
+I0922 23:28:40.352017 1 http_server.cc:3558] Started HTTPService at 0.0.0.0:8000
+I0922 23:28:40.395611 1 http_server.cc:187] Started Metrics Service at 0.0.0.0:8002
+```
+
+## Step 4: Use a Triton Client to Query the Server
```
Comment (rmccorm4) [inline Quick_Deploy/HuggingFaceTransformers/README.md:91]:
```suggestion
## Step 4: Query the Server
```

File: Quick_Deploy/HuggingFaceTransformers/README.md [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+
+# persimmon8b
+"Where is the nearest starbucks?"
+```
+## 'Hour Zero' Support
+
+At the time of writing this tutorial, transformers version 4.34.0 was not yet released, meaning
+very new models such as Persimmon-8B and Mistral 7B were not yet fully supported by the latest
+transformers releases (4.33.3). Triton is not limited to waiting for official releases and can
+load the latest models by building from source. This can be done by replacing:
```
Comment (rmccorm4) [inline Quick_Deploy/HuggingFaceTransformers/README.md:158]: We may want to avoid specific versions and models here to avoid dating it too much, as these models mentioned will quickly become "old" compared to newer ones. We should instead just say something more generic about how bleeding edge / latest models may not be supported in the latest published release of `transformers`, so for bleeding edge model support you can install latest transformers from source. To do so, replace the transformers install in the `Dockerfile` with `pip install git+https://github.com/huggingface/transformers.git` ...

Using this technique you should be able to serve any models supported by HuggingFace Transformer text generation pipeline with Triton.

---

above is just the gist, feel free to rephrase, just want to avoid mention of specific versions and models in this disclaimer

File: Quick_Deploy/HuggingFaceTransformers/README.md (hunk 11 above)
Comment (fpetrini15) [reply Quick_Deploy/HuggingFaceTransformers/README.md:42]: I agree. In that same vein then do we want to change the the WIP phrase for the vLLM [tutorial](https://github.com/triton-inference-server/tutorials/blob/512ba538fd81b37ad0cc3133a6d7905f72a9ad04/Quick_Deploy/vLLM/README.md?plain=1#L37C66-L37C91) to something like: 
"The tutorial is intended to be a reference example only and has known limitations."?

File: Quick_Deploy/HuggingFaceTransformers/README.md (hunk 12 above)
Comment (fpetrini15) [reply Quick_Deploy/HuggingFaceTransformers/README.md:49]: TIL

File: Quick_Deploy/HuggingFaceTransformers/README.md (hunk 14 above)
Comment (fpetrini15) [reply Quick_Deploy/HuggingFaceTransformers/README.md:158]: Great point!

File: Quick_Deploy/HuggingFaceTransformers/README.md (hunk 11 above)
Comment (rmccorm4) [reply Quick_Deploy/HuggingFaceTransformers/README.md:42]: sounds good to me CC @tanmayv25

**Review by rmccorm4 — COMMENTED** [review]
Great tutorial overall! Only minor comments 🚀

File: Quick_Deploy/HuggingFaceTransformers/README.md (hunk 9 above)
Comment (fpetrini15) [reply Quick_Deploy/HuggingFaceTransformers/README.md]: I appreciated the cheekiness, but I guess we can change it to something more tame... :disappointed:

File: Quick_Deploy/HuggingFaceTransformers/falcon7b/1/model.py [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+        self.logger.log_info(f"Loading HuggingFace model: {hf_model}...")
+        # Assume tokenizer available for same model
+        self.tokenizer = transformers.AutoTokenizer.from_pretrained(hf_model)
+        self.pipeline = transformers.pipeline(
+            "text-generation",
+            model=hf_model,
+            torch_dtype=torch.float16,
+            tokenizer=self.tokenizer,
+            device_map="auto",
+            trust_remote_code=True,
```
Comment (rmccorm4) [inline Quick_Deploy/HuggingFaceTransformers/falcon7b/1/model.py:64]: Why is this needed for falcon?

Comment (fpetrini15) [reply Quick_Deploy/HuggingFaceTransformers/falcon7b/1/model.py:64]: When I first was testing with falcon weeks ago I believe it was necessary, but looks like [recent changes](https://huggingface.co/tiiuae/falcon-7b/commit/898df1396f35e447d5fe44e0a3ccaaaa69f30d36) in the past week have made it fully supported by the transformers library. I tested the tutorial without this parameter and it works fine so I'll remove it.

File: Quick_Deploy/HuggingFaceTransformers/README.md (hunk 11 above)
Comment (fpetrini15) [reply Quick_Deploy/HuggingFaceTransformers/README.md]: I went ahead and changed it. I can revert it if need be.

**Comment by fpetrini15** [comment]
@nnshah1. I preemptively removed Mistral from the tutorial. I can always revert if necessary.

**Comment by fpetrini15** [comment]
Incorporated some feedback from Dora incorporating how to gather performance metrics, load cached models, and adding comments.

File: Quick_Deploy/HuggingFaceTransformers/persimmon8b/config.pbtxt [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+	key: "max_output_length",
+	value: {string_value: "15"}
+}
+
+# Triton should expect a single
+# string of set length named prompt
+# as input
+input [
+  {
+    name: "prompt"
```
Comment (nnshah1) [inline Quick_Deploy/HuggingFaceTransformers/persimmon8b/config.pbtxt:25]: could we update these with "text-input" and "text-output" as the names to match up with vLLM backend? Not critical but nice to have.

File: Quick_Deploy/HuggingFaceTransformers/falcon7b/config.pbtxt [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+	key: "max_output_length",
+	value: {string_value: "15"}
+}
+
+# Triton should expect a single
+# string of set length named prompt
+# as input
+input [
+  {
+    name: "prompt"
```
Comment (nnshah1) [inline Quick_Deploy/HuggingFaceTransformers/falcon7b/config.pbtxt:25]: can we use "text-input" and "text-output" to match up with vLLM backend?

File: Quick_Deploy/HuggingFaceTransformers/persimmon8b/config.pbtxt (hunk 16 above)
Comment (rmccorm4) [reply Quick_Deploy/HuggingFaceTransformers/persimmon8b/config.pbtxt:25]: +1 on renaming

Also, since there's no limitations like optional inputs or decoupled mode being used here, we can define the I/O names in auto complete function instead of config file.

Comment (fpetrini15) [reply Quick_Deploy/HuggingFaceTransformers/persimmon8b/config.pbtxt:25]: The tutorial initially used auto-complete for the input and outputs, but in an [earlier review ](https://github.com/triton-inference-server/tutorials/pull/49#discussion_r1335027432)it was decided to add this information to the config.pbtxt files instead.

File: Quick_Deploy/HuggingFaceTransformers/README.md [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+increasing levels of concurrency. As the analyzer runs, it will output useful metrics
+such as latency percentiles, latency by stage of inference, and successful request
+count. Ultimately, the analyzer will neatly summarize the data in the final output:
+
+```json
+Concurrency: 1, throughput: 23.2174 infer/sec, latency 43041 usec
+Concurrency: 2, throughput: 23.3284 infer/sec, latency 85590 usec
+Concurrency: 3, throughput: 23.94 infer/sec, latency 125085 usec
+Concurrency: 4, throughput: 23.773 infer/sec, latency 167879 usec
+```
```
Comment (rmccorm4) [inline Quick_Deploy/HuggingFaceTransformers/README.md:218]: Not sure if this is specific feedback, but since the models aren't optimized for batching, the throughput stays flat here and latency gets worse linearly with concurrency - so I don't think we should show concurrency > 1 for now.

File: Quick_Deploy/HuggingFaceTransformers/falcon7b/config.pbtxt [hunk 19]
```diff
@@ -0,0 +1,40 @@
+# Triton backend to use
+backend: "python"
+
+# Hugging face model path
+# Parameters must follow this
+# key/value structure
```
Comment (rmccorm4) [inline Quick_Deploy/HuggingFaceTransformers/falcon7b/config.pbtxt:6]: The formatting is a bit weird/inconsistent in this file:
1. Comments are weirdly narrow - can probably fit on 1 or 2 lines
2. Parameters a different indentation than I/O

File: Quick_Deploy/HuggingFaceTransformers/persimmon8b/config.pbtxt (hunk 16 above)
Comment (rmccorm4) [reply Quick_Deploy/HuggingFaceTransformers/persimmon8b/config.pbtxt]: Oops, OK let's keep it in config for tutorial for now then, sorry for the confusion.

Comment (fpetrini15) [reply Quick_Deploy/HuggingFaceTransformers/persimmon8b/config.pbtxt]: No worries!

File: Quick_Deploy/HuggingFaceTransformers/falcon7b/config.pbtxt (hunk 19 above)
Comment (fpetrini15) [reply Quick_Deploy/HuggingFaceTransformers/falcon7b/config.pbtxt]: 1. I was trying to pad my contributed lines stat 😉. Jk fixed,
2. Fixed.

File: Quick_Deploy/HuggingFaceTransformers/README.md (hunk 18 above)
Comment (fpetrini15) [reply Quick_Deploy/HuggingFaceTransformers/README.md]: We were asked to incorporate tangible ways to measure performance and optimize a generic deployment. Seemed like a great time to point to the tools team. But yes, this is a strong point! For now, I've changed it to just collect metrics with concurrency 1. The full detail list the perf analyzer produces is this:
```Request concurrency: 1
  Client:
    Request count: 422
    Throughput: 23.4404 infer/sec
    Avg latency: 42583 usec (standard deviation 3809 usec)
    p50 latency: 41547 usec
    p90 latency: 43611 usec
    p95 latency: 48919 usec
    p99 latency: 61778 usec
    Avg HTTP time: 42576 usec (send/recv 87 usec + response wait 42489 usec)
  Server:
    Inference count: 422
    Execution count: 422
    Successful request count: 422
    Avg request latency: 42140 usec (overhead 24 usec + queue 26 usec + compute input 26 usec + compute infer 41998 usec + compute output 65 usec)

  Server Prometheus Metrics:
    Avg GPU Utilization:
      GPU-57c7b00e-ca04-3876-91e2-c1eae40a0733 : 52.7222%
    Avg GPU Power Usage:
      GPU-57c7b00e-ca04-3876-91e2-c1eae40a0733 : 234.525 watts
    Max GPU Memory Usage:
      GPU-57c7b00e-ca04-3876-91e2-c1eae40a0733 : 35431383040 bytes
    Total GPU Memory:
      GPU-57c7b00e-ca04-3876-91e2-c1eae40a0733 : 51527024640 bytes
Inferences/Second vs. Client Average Batch Latency
Concurrency: 1, throughput: 23.4404 infer/sec, latency 42583 usec```

If you think any of these metrics would be more relevant than the ones I selected, let me know and I can swap them out!

Comment (rmccorm4) [reply Quick_Deploy/HuggingFaceTransformers/README.md]: Looks great! I haven't actually seen the GPU metric collection feature being used until now, very cool!

Comment (rmccorm4) [reply Quick_Deploy/HuggingFaceTransformers/README.md]: If you have spare cycles for a couple more runs - I notice the GPU utilization is ~50%. I wonder if we can hit 100% with 2+ instances of the model (if it can fit, I see memory usage is > half).

Comment (rmccorm4) [reply Quick_Deploy/HuggingFaceTransformers/README.md]: (Disclaimer) This and the previous comment are totally separate from this tutorial PR and only if you have spare cycles/curiosity.

If simply scaling instances doesn't work, another thing to try eventually is batching. Long term we probably want to move away from pipeline to have more fine tuned control over batching, but for a quick example, you could enable dynamic batching on the model, either set high concurrency on perf analyzer or set high max_queue_delay in config.pbtxt to allow larger batch to be gathered, and I think you can do something like this to [batch with the pipeline](https://huggingface.co/docs/transformers/main_classes/pipelines#pipeline-batching):
```python
prompts = []
for request in requsts:
   prompts.append(prompt)

# or some fixed batch size? 
batch_size = len(prompts)
sequences = self.pipeline(prompts, batch_size=batch_size, ...)

responses = []
for seq in sequences:
  # tensor = ...
  # response = ...
  responses.append(response)

return responses
```

Comment (fpetrini15) [reply Quick_Deploy/HuggingFaceTransformers/README.md]: As a matter of fact, I know it can hit ~100%, because when I was building out the model analyzer part of the tutorial it recommended a config with 2 GPU instances for that very reason. Maybe better to show the output in the MA section to show this very value.

File: Quick_Deploy/HuggingFaceTransformers/README.md [hunk 20]
```diff
@@ -0,0 +1,263 @@
...
+I0922 23:28:40.352017 1 http_server.cc:3558] Started HTTPService at 0.0.0.0:8000
+I0922 23:28:40.395611 1 http_server.cc:187] Started Metrics Service at 0.0.0.0:8002
+```
+
+## Step 4: Query the Server
+
+Now we can query the server using curl, specifying the server address and input details:
+
+```json
+curl -X POST localhost:8000/v2/models/falcon7b/infer -d '{"inputs": [{"name":"text_input","datatype":"BYTES","shape":[1],"data":["I am going"]}]}'
```
Comment (rmccorm4) [inline Quick_Deploy/HuggingFaceTransformers/README.md:97]: We can also curl with the `/generate` endpoint now. Feel free to add in a separate PR if you want though.

Comment (rmccorm4) [reply Quick_Deploy/HuggingFaceTransformers/README.md:97]: ref: https://github.com/triton-inference-server/server/blob/main/docs/protocol/extension_generate.md

File: Quick_Deploy/HuggingFaceTransformers/README.md [hunk 21]
```diff
@@ -0,0 +1,263 @@
...
+is downloaded, we can mount it to the Triton container by adding the following mount option to our
+`docker run` command from earlier (making sure to replace {USER} with your username on
+Linux):
+
+```bash
+# Option to mount a specific cached model (falcon-7b in this case)
+-v /home/{USER}/.cache/huggingface/hub/models--tiiuae--falcon-7b:/root/.cache/huggingface/hub/models--tiiuae--falcon-7b
+
+# Option to mount all cached models on the host system
+-v /home/{USER}/.cache/huggingface:/root/.cache/huggingface
```
Comment (rmccorm4) [inline Quick_Deploy/HuggingFaceTransformers/README.md:182]: This is a great addition! Do you know if the cache directory is configurable for HF? Would be cool if we could eventually set it to cache where the model.py lives / in the model repository or something.

Comment (rmccorm4) [inline Quick_Deploy/HuggingFaceTransformers/README.md:182]: `/home/${USER}` can be `${HOME}` instead for brevity

Comment (fpetrini15) [reply Quick_Deploy/HuggingFaceTransformers/README.md:182]: Indeed it is, and indeed we can. I'll extend this section.

File: Quick_Deploy/HuggingFaceTransformers/README.md (hunk 20 above)
Comment (fpetrini15) [reply Quick_Deploy/HuggingFaceTransformers/README.md:97]: I'll create a minor ticket for this and wait for the 23.10 container to be released.

**Comment by fpetrini15** [comment]
CC @nv-braf @matthewkotila in case there is any feedback regarding the PA/MA section.

**Review by jbkyang-nvi — APPROVED** [review]
